src/middlewared/middlewared/plugins/kerberos.py

```diff
@@ -1028,18 +1031,27 @@ async def _prune_keytab_principals(self, to_delete=[]):
                 seen_principals.append(i['principal'])
 
         else:
+            """
+            The pruned keytab must be written to a new file to avoid duplication of
+            entries.
+            """
             rkt = f"rkt {keytab.SAMBA.value}"
-            wkt = f"wkt {keytab.SAMBA.value}"
-            for i in reversed(to_delete):
-                ktutil_remove = await Popen(['ktutil'],
-                                            stdout=subprocess.PIPE,
-                                            stderr=subprocess.PIPE,
-                                            stdin=subprocess.PIPE)
-                output = await ktutil_remove.communicate(
-                    f'{rkt}\ndelent {i["slot"]}\n{wkt}'
-                )
-                if output[1]:
-                    raise CallError(output[1].decode())
+            wkt = "wkt /var/db/system/samba4/samba_mit.keytab"
+            delents = "\n".join(f"delent {x['slot']}" for x in reversed(to_delete))
+            ktutil_remove = await Popen(['ktutil'],
+                                        stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE,
+                                        stdin=subprocess.PIPE)
+            output = await ktutil_remove.communicate(
+                f'{rkt}\n{delents}\n{wkt}\nq\n'.encode()
+            )
+            if output[1]:
+                raise CallError(output[1].decode())
+
+            with contextlib.suppress(OSError):
+                os.remove(keytab.SAMBA.value)
+
+            os.rename("/var/db/system/samba4/samba_mit.keytab", keytab.SAMBA.value)
```

Comment:
This will fail if `"/var/db/system/samba4/samba_mit.keytab"` and `keytab.SAMBA.value` are on different devices. `shutil.move` is a good replacement.

Reply:
Or just set `wkt = f"wkt {keytab.SAMBA.value}.tmp.keytab"`